mesonbuild/compilers/cuda.py

```diff
@@ -384,16 +383,19 @@ def _to_host_flags(cls, flags: T.List[str], phase: _Phase = _Phase.COMPILER) ->
                 else:                            # -Isomething
                     val = flag[2:]
                 flag = flag[:2]                  # -I
-            elif flag in cls._FLAG_LONG2SHORT_WITHARGS or \
-                 flag in cls._FLAG_SHORT2LONG_WITHARGS:
+            elif flag.startswith('-include') or flag.startswith('-isystem'):
+                val = flag[8:].strip()
+                flag = flag[:8]                  # -include/-isystem
```

Comment:
Trying to read this, I found it hard to realize you're special casing this because `-isystem /usr/include` is handled by "normal option parsing", but the quite odd variant `-isystem/usr/include` wouldn't be detected properly.

Could you comment this? The current comment is really quite useless, the startswith check tells us the "what" already, but an actual description of "why" would be nice.

Reply:
both `-isystem /usr/include` and `-isystem/usr/include` get caught by this code.

So you want me to comment here instead of later? I have a comment for treating `-isystem` and catching `/usr/include` later when the actual processing/filtering happens

Reply:
@SoapGentoo @eli-schwartz I am the author of this file and of the arguments parsing, for good or ill.

I believe I understand what my screw-up was here. The difference appears to be that in NVCC, `-isystem` is a list option and is documented as requiring that it be separated from its value by either a space or an `=` sign. Therefore, `['-isystemfoobar']` (perhaps constructed from string concatenation) is an illegal argument. However, GCC appears to accept it just fine (!) and interprets it the same as `-isystem foobar`. The relative path `foobar` doesn't even need to exist apparently. I falsely assumed GCC would reject it as well for similar reasons due to potential ambiguity (in its `man`page, the two following flags are `-iwithprefix dir` and `-iwithprefixbefore dir`, which exhibit precisely that).

I think there is something subtly wrong in the patch as proposed. Because this special-case is found *before* the `elif flag in self._FLAG_LONG2SHORT_WITHARGS or flag in self._FLAG_SHORT2LONG_WITHARGS:` test, it will capture `-isystem` and `-include` with value concatenated (`-isystemfoobar`), separated (`-isystem=foobar`) *and* isolated (`-isystem foobar`), but in the latter case it will compute `'-isystem'[8:] == ''` as the value and then consider `foobar` a standalone argument.

If this could be fixed and a comment explaining the foregoing added, I'd be very happy.

Reply:
Ah yes, and to make things even more complicated, it turns out that GCC considers that if the first character of the directory argument of `-I -iquote -isystem -dirafter` is `=`, then the path is assumed to be relative to the sysroot prefix. That introduces an ambiguity I did not heretofore think about when writing this code, as it is impossible for me to guess what `-I=foobar` should be understood as (GCC-style sysroot include, or NVCC-style regular include?). How painful.

```
       -I dir
       -iquote dir
       -isystem dir
       -idirafter dir
           Add the directory dir to the list of directories to be searched for header files during preprocessing.  If dir
           begins with = or $SYSROOT, then the = or $SYSROOT is replaced by the sysroot prefix; see --sysroot and -isysroot.
```

Reply:
I'd assume if nvcc successfully parses it as an option at all, it reinterprets it... so err on the side of NVCC style parsing.

And yes, I don't know why GCC supports it without spaces but if you look through the meson codebase you'll see we use both forms.

(e.g. the generic clike mixin defines get_include_args as `['-isystem', '/path']` while gnu defines it to `['-isystem/path']`)

Reply:
@obilaniu are you on IRC? discussing this on IRC seems a lot more promising than here.

Reply:
@SoapGentoo Ay, no I have no account there. Anything in particular you wanted to ask about?